CTFd/__init__.py

```diff
@@ -100,35 +100,31 @@ def _load_template(self, name, globals):
 
 
 class ThemeLoader(FileSystemLoader):
-    """Custom FileSystemLoader that switches themes based on the configuration value"""
-
-    def __init__(self, searchpath, encoding="utf-8", followlinks=False):
+    """Custom FileSystemLoader that is aware of theme structure and config.
+    """
+
+    DEFAULT_THEMES_PATH = os.path.join(os.path.dirname(__file__), "themes")
+    _ADMIN_THEME_PREFIX = ADMIN_THEME + "/"
+
+    def __init__(
+        self,
+        searchpath=DEFAULT_THEMES_PATH,
+        theme_name=None,
+        encoding="utf-8",
+        followlinks=False,
+    ):
         super(ThemeLoader, self).__init__(searchpath, encoding, followlinks)
-        self.overriden_templates = {}
+        self.theme_name = theme_name
 
     def get_source(self, environment, template):
-        # Check if the template has been overriden
-        if template in self.overriden_templates:
-            return self.overriden_templates[template], template, lambda: True
-
-        # Check if the template requested is for the admin panel
-        if template.startswith("admin/"):
-            template = template[6:]  # Strip out admin/
-            template = "/".join(["admin", "templates", template])
-            return super(ThemeLoader, self).get_source(environment, template)
-
-        # Load regular theme data with potential fallbacks
-        first_cand, *others = ctf_theme_candidates()
-        for cand_theme in (first_cand, *others):
-            tpl_path = safe_join(cand_theme, "templates", template)
-            try:
-                return super(ThemeLoader, self).get_source(environment, tpl_path)
-            except jinja2.exceptions.TemplateNotFound:
-                pass
-        else:
-            # If we found no matching templates then raise an exception for the
-            # first one we intended to try
-            raise jinja2.exceptions.TemplateNotFound(first_cand)
+        # Refuse to load `admin/*` from a loader not for the admin theme
+        if template.startswith(self._ADMIN_THEME_PREFIX):
+            if self.theme_name != ADMIN_THEME:
+                raise jinja2.TemplateNotFound(template)
+            template = template[len(self._ADMIN_THEME_PREFIX) :]
```

Comment:
This seems to differ from the previous code but I can't tell why it's needed. 

Wouldn't the admin themes only ever be handled by the admin ThemeLoader?

Reply:
This is intended as a bit of a mitigation against evil themes. The idea being that if an admin installs a theme which provides a `templates/admin/set_password.html` file, they might get tricked into entering data on a malicious form. It's not particularly strong and perhaps a better approach would be to move the prefix loader (which includes the admin theme) in front of the default loader when we construct the `ChoiceLoader`. That would mean the admin theme would always get first chance at resolving `admin/...` template paths and as long as it doesn't refer to absent template files, the default theme wouldn't be able to add it's own content to the admin panels.

In any case, this could use a comment to explain itself if we choose to leave it as is.

Thoughts?

Reply:
Moving the preflix loader in front of the default loader would have the consequence that collisions between the name of another theme (say `foo`) and a subdirectory of the default theme's templates (say `bar/templates/foo`) could end up rendering a template from `foo` rather than the expected one from `bar`.

Reply:
I see the concern now after testing it out. Kind of stems from the lack of separate template environments in Jinja I think but I think your solution is good.